server/proto/merger.go

```diff
@@ -0,0 +1,228 @@
+		if getFileDescriptorProto(prevRegistry, currFileDesc.Path()) == nil {
+			mergedFiles = append(mergedFiles, currFile)
+		}
+		return true
+	})
+	return mergedFiles
+}
+
+// add new message from currMsgs to prevMsgs
+func addMessage(prevMsgs *[]*descriptorpb.DescriptorProto, currMsgs *[]*descriptorpb.DescriptorProto) {
+	for _, curr := range *currMsgs {
+		exist := false
+		for _, prev := range *prevMsgs {
+			if prev.GetName() == curr.GetName() {
+				exist = true
+				break
+			}
+		}
+		if !exist {
+			*prevMsgs = append(*prevMsgs, curr)
+		}
+	}
+}
+
+// Append all new message in current file to previous file.
+func addNewMessage(prevFile *descriptorpb.FileDescriptorProto, currFile *descriptorpb.FileDescriptorProto) {
+	addMessage(&prevFile.MessageType, &currFile.MessageType)
+}
+
+// Append all new nested message in current message to previous message.
+func addNewNestedMessage(prevParent *descriptorpb.DescriptorProto, currParent *descriptorpb.DescriptorProto) {
+	addMessage(&prevParent.NestedType, &currParent.NestedType)
+}
+
+// Add all new import statements in current file to previous file.
+func addNewDependencies(prevFile *descriptorpb.FileDescriptorProto, currFile *descriptorpb.FileDescriptorProto) {
+	for _, currDep := range currFile.GetDependency() {
+		isNew := true
+		for _, prevDep := range prevFile.GetDependency() {
+			if currDep == prevDep {
+				isNew = false
+				break
+			}
+		}
+		if isNew {
+			prevFile.Dependency = append(prevFile.Dependency, currDep)
+		}
+	}
+}
+
+// Set deprecated options to true for each removed field in message.
+func deprecateRemovedFields(prev *descriptorpb.DescriptorProto, curr *descriptorpb.DescriptorProto) {
+	for _, prevField := range prev.Field {
+		removed := true
+		for _, currField := range curr.Field {
+			if prevField.GetName() == currField.GetName() {
+				removed = false
+				break
+			}
+		}
+		if removed {
+			deprecated := true
+			prevField.Options = &descriptorpb.FieldOptions{Deprecated: &deprecated}
+		}
+	}
+}
+
+// New field is added after the max number of existing fields.
+func addNewFields(prev *descriptorpb.DescriptorProto, curr *descriptorpb.DescriptorProto) {
+	// get max number in previous message
+	var maxPrevNumber int32 = 0
+	for _, prevField := range prev.Field {
+		if *prevField.Number > maxPrevNumber {
+			maxPrevNumber = *prevField.Number
+		}
+	}
+	// add new fields after max number
+	for _, currField := range curr.Field {
+		isNew := true
+		for _, prevField := range prev.Field {
+			if currField.GetName() == prevField.GetName() {
+				isNew = false
+				break
+			}
+		}
+		if isNew {
+			fieldNumber := maxPrevNumber + 1
+			currField.Number = &fieldNumber
+			prev.Field = append(prev.Field, currField)
+			maxPrevNumber = maxPrevNumber + 1
+		}
+	}
+}
+
+// Set deprecated options of message to true.
+func deprecateMessage(msg *descriptorpb.DescriptorProto) {
+	deprecated := true
+	msg.Options = &descriptorpb.MessageOptions{Deprecated: &deprecated}
```

Comment:
retain existing options data if exists then add deprecated option. This will overwrite any existing message options.